# Fifa practice

We want to predict the *value* of the players using the information of the player.

## Import libraries

First of all we import the libraries we're gonna use.

In [1]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

## Read the data

To read the data we're gonna use ``pandas`` and an utility from the library
``os``.


In [2]:
df = pd.read_csv(os.path.join("..", "in", "fifa.csv"))
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]

Now we select one columns. The result will be an object from the ``pandas`` library a ``series`` object. Thi object can be converted to a more standard ``numpy`` array.

## Data analysis 

First of all we want to known what is happening to the data. To do so we increase the number of columns to show. After that we're gonna use a special instruction that will show us the data structure.



In [3]:
pd.set_option('display.max_columns', None)

In [4]:
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

In [5]:
df.describe()

Unnamed: 0             ID           Age       Overall     Potential  \
count  18207.000000   18207.000000  18207.000000  18207.000000  18207.000000   
mean    9103.000000  214298.338606     25.122206     66.238699     71.307299   
std     5256.052511   29965.244204      4.669943      6.908930      6.136496   
min        0.000000      16.000000     16.000000     46.000000     48.000000   
25%     4551.500000  200315.500000     21.000000     62.000000     67.000000   
50%     9103.000000  221759.000000     25.000000     66.000000     71.000000   
75%    13654.500000  236529.500000     28.000000     71.000000     75.000000   
max    18206.000000  246620.000000     45.000000     94.000000     95.000000   

            Special  International Reputation     Weak Foot   Skill Moves  \
count  18207.000000              18159.000000  18159.000000  18159.000000   
mean    1597.809908                  1.113222      2.947299      2.361308   
std      272.586016                  0.394031      0.660456      0.756164   
min      731.000000                  1.000000      1.000000      1.000000   
25%     1457.000000                  1.000000      3.000000      2.000000   
50%     1635.000000                  1.000000      3.000000      2.000000   
75%     1787.000000                  1.000000      3.000000      3.000000   
max     2346.000000                  5.000000      5.000000      5.000000   

       Jersey Number      Crossing     Finishing  HeadingAccuracy  \
count   18147.000000  18159.000000  18159.000000     18159.000000   
mean       19.546096     49.734181     45.550911        52.298144   
std        15.947765     18.364524     19.525820        17.379909   
min         1.000000      5.000000      2.000000         4.000000   
25%         8.000000     38.000000     30.000000        44.000000   
50%        17.000000     54.000000     49.000000        56.000000   
75%        26.000000     64.000000     62.000000        64.000000   
max        99.000000     93.000000     95.000000        94.000000   

       ShortPassing       Volleys     Dribbling         Curve    FKAccuracy  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      58.686712     42.909026     55.371001     47.170824     42.863153   
std       14.699495     17.694408     18.910371     18.395264     17.478763   
min        7.000000      4.000000      4.000000      6.000000      3.000000   
25%       54.000000     30.000000     49.000000     34.000000     31.000000   
50%       62.000000     44.000000     61.000000     48.000000     41.000000   
75%       68.000000     57.000000     68.000000     62.000000     57.000000   
max       93.000000     90.000000     97.000000     94.000000     94.000000   

        LongPassing   BallControl  Acceleration   SprintSpeed       Agility  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      52.711933     58.369459     64.614076     64.726967     63.503607   
std       15.327870     16.686595     14.927780     14.649953     14.766049   
min        9.000000      5.000000     12.000000     12.000000     14.000000   
25%       43.000000     54.000000     57.000000     57.000000     55.000000   
50%       56.000000     63.000000     67.000000     67.000000     66.000000   
75%       64.000000     69.000000     75.000000     75.000000     74.000000   
max       93.000000     96.000000     97.000000     96.000000     96.000000   

          Reactions       Balance     ShotPower       Jumping       Stamina  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      61.836610     63.966573     55.460047     65.089432     63.219946   
std        9.010464     14.136166     17.237958     11.820044     15.894741   
min       21.000000     16.000000      2.000000     15.000000     12.000000   
25%       56.000000     56.000000     45.000000     58.000000     56.000000   
50%       62.000000     66.000000     59.000000     66.000000     66.00000

We can see that the two first columns do not provide any meaningful information so we can delete them.

In [6]:
df = df.iloc[:, 2:]
df

Name  Age  \
0                L. Messi   31   
1       Cristiano Ronaldo   33   
2               Neymar Jr   26   
3                  De Gea   27   
4            K. De Bruyne   27   
...                   ...  ...   
18202        J. Lundstram   19   
18203  N. Christoffersson   19   
18204           B. Worman   16   
18205      D. Walker-Rice   17   
18206           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate   Body Type Real Face Position  \
0            4.0          4.0  Medium/ Medium       Messi       Yes     

## DELETE COLUMNS

Eliminamos las columnas que consideramos que no son importantes para la predicción del precio

In [7]:
df = df.drop(["Contract Valid Until","Photo","Flag","Club Logo","Preferred Foot","Nationality","Weak Foot","Body Type","Real Face","Jersey Number","Name","Loaned From","Work Rate","Joined"], axis=1)

df

Age  Overall  Potential                 Club    Value   Wage  Special  \
0       31       94         94         FC Barcelona  €110.5M  €565K     2202   
1       33       94         94             Juventus     €77M  €405K     2228   
2       26       92         93  Paris Saint-Germain  €118.5M  €290K     2143   
3       27       91         93    Manchester United     €72M  €260K     1471   
4       27       91         92      Manchester City    €102M  €355K     2281   
...    ...      ...        ...                  ...      ...    ...      ...   
18202   19       47         65      Crewe Alexandra     €60K    €1K     1307   
18203   19       47         63       Trelleborgs FF     €60K    €1K     1098   
18204   16       47         67     Cambridge United     €60K    €1K     1189   
18205   17       47         66      Tranmere Rovers     €60K    €1K     1228   
18206   16       46         66      Tranmere Rovers     €60K    €1K     1321   

       International Reputation  Skill Moves Position Height  Weight    LS  \
0                           5.0          4.0       RF    5'7  159lbs  88+2   
1                           5.0          5.0       ST    6'2  183lbs  91+3   
2                           5.0          5.0       LW    5'9  150lbs  84+3   
3                           4.0          1.0       GK    6'4  168lbs   NaN   
4                           4.0          4.0      RCM   5'11  154lbs  82+3   
...                         ...          ...      ...    ...     ...   ...   
18202                       1.0          2.0       CM    5'9  134lbs  42+2   
18203                       1.0          2.0       ST    6'3  170lbs  45+2   
18204                       1.0          2.0       ST    5'8  148lbs  45+2   
18205                       1.0          2.0       RW   5'10  154lbs  47+2   
18206                       1.0          2.0       CM   5'10  176lbs  43+2   

         ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0      88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1      91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2      84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
18203  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2   
18204  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2   
18205  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2   
18206  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2   

         CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0      84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1      81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2      81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2  45+2   
18203  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2  32+2  32+2  32+2   
18204  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2  28+2  28+2  28+2   
18205  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2  31+2  31+2  31+2   
18206  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2  47+2  47+2  47+2   

         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0      59+2      84.0       95.0             70.0          90.0     86.0   
1      61+3

Buscamos los jugadores que juegan cómo porteros para valorar sus datos

In [8]:
df.loc[df.Position == "GK"]

Age  Overall  Potential               Club   Value   Wage  Special  \
3       27       91         93  Manchester United    €72M  €260K     1471   
9       25       90         93    Atlético Madrid    €68M   €94K     1331   
18      26       89         92       FC Barcelona    €58M  €240K     1328   
19      26       89         90        Real Madrid  €53.5M  €240K     1311   
22      32       89         89  FC Bayern München    €38M  €130K     1473   
...    ...      ...        ...                ...     ...    ...      ...   
18178   18       48         65         Dalkurd FF    €50K    €1K      738   
18180   22       48         58   St. Johnstone FC    €40K    €1K      987   
18183   44       48         48   Cambridge United      €0    €1K      774   
18194   18       47         65              Lecce    €50K    €1K      731   
18198   18       47         70      Burton Albion    €60K    €1K      792   

       International Reputation  Skill Moves Position Height  Weight   LS  \
3                           4.0          1.0       GK    6'4  168lbs  NaN   
9                           3.0          1.0       GK    6'2  192lbs  NaN   
18                          3.0          1.0       GK    6'2  187lbs  NaN   
19                          4.0          1.0       GK    6'6  212lbs  NaN   
22                          5.0          1.0       GK    6'4  203lbs  NaN   
...                         ...          ...      ...    ...     ...  ...   
18178                       1.0          1.0       GK    6'0  176lbs  NaN   
18180                       1.0          1.0       GK    6'1  172lbs  NaN   
18183                       1.0          1.0       GK    6'2  190lbs  NaN   
18194                       1.0          1.0       GK    6'3  187lbs  NaN   
18198                       1.0          1.0       GK   5'11  154lbs  NaN   

        ST   RS   LW   LF   CF   RF   RW  LAM  CAM  RAM   LM  LCM   CM  RCM  \
3      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
19     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
22     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
18178  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18180  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18183  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18194  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18198  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

        RM  LWB  LDM  CDM  RDM  RWB   LB  LCB   CB  RCB   RB  Crossing  \
3      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      17.0   
9      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      13.0   
18     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      15.0   
19     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      14.0   
22     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      15.0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...       ...   
18178  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      10.0   
18180  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      19.0   
18183  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      11.0   
18194  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      10.0   
18198  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      14.0   

       Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
3           13.0             21.0          50.0     13.0       18.0   21.0   
9           11.0             15.0          29.0     13.0       12.0   13.0   
18          14.0             11.0          36.0     14.0       17.0   18.0   
19     

Cómo podemos ver, los porteros tienen una gran parte de valores *NaN*, por lo que para nuestro entrenamiento para hallar el valor, no nos será de gran ayuda y decidimos eliminarlo

In [9]:
df = df.drop(df[df['Position']== 'GK'].index)
df

Age  Overall  Potential                 Club    Value   Wage  Special  \
0       31       94         94         FC Barcelona  €110.5M  €565K     2202   
1       33       94         94             Juventus     €77M  €405K     2228   
2       26       92         93  Paris Saint-Germain  €118.5M  €290K     2143   
4       27       91         92      Manchester City    €102M  €355K     2281   
5       27       91         91              Chelsea     €93M  €340K     2142   
...    ...      ...        ...                  ...      ...    ...      ...   
18202   19       47         65      Crewe Alexandra     €60K    €1K     1307   
18203   19       47         63       Trelleborgs FF     €60K    €1K     1098   
18204   16       47         67     Cambridge United     €60K    €1K     1189   
18205   17       47         66      Tranmere Rovers     €60K    €1K     1228   
18206   16       46         66      Tranmere Rovers     €60K    €1K     1321   

       International Reputation  Skill Moves Position Height  Weight    LS  \
0                           5.0          4.0       RF    5'7  159lbs  88+2   
1                           5.0          5.0       ST    6'2  183lbs  91+3   
2                           5.0          5.0       LW    5'9  150lbs  84+3   
4                           4.0          4.0      RCM   5'11  154lbs  82+3   
5                           4.0          4.0       LF    5'8  163lbs  83+3   
...                         ...          ...      ...    ...     ...   ...   
18202                       1.0          2.0       CM    5'9  134lbs  42+2   
18203                       1.0          2.0       ST    6'3  170lbs  45+2   
18204                       1.0          2.0       ST    5'8  148lbs  45+2   
18205                       1.0          2.0       RW   5'10  154lbs  47+2   
18206                       1.0          2.0       CM   5'10  176lbs  43+2   

         ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0      88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1      91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2      84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
4      82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   
5      83+3  83+3  89+3  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
18203  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2   
18204  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2   
18205  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2   
18206  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2   

         CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0      84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1      81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2      81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
4      87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   
5      82+3  82+3  89+3  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2  45+2   
18203  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2  32+2  32+2  32+2   
18204  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2  28+2  28+2  28+2   
18205  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2  31+2  31+2  31+2   
18206  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2  47+2  47+2  47+2   

         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0      59+2      84.0       95.0             70.0          90.0     86.0   
1      61+3

Después de haber eliminado a los porteros, reseteamos los indices para que vuelvan a tener un orden correcto

In [10]:
df = df.reset_index(drop=True)
df

Age  Overall  Potential                 Club    Value   Wage  Special  \
0       31       94         94         FC Barcelona  €110.5M  €565K     2202   
1       33       94         94             Juventus     €77M  €405K     2228   
2       26       92         93  Paris Saint-Germain  €118.5M  €290K     2143   
3       27       91         92      Manchester City    €102M  €355K     2281   
4       27       91         91              Chelsea     €93M  €340K     2142   
...    ...      ...        ...                  ...      ...    ...      ...   
16177   19       47         65      Crewe Alexandra     €60K    €1K     1307   
16178   19       47         63       Trelleborgs FF     €60K    €1K     1098   
16179   16       47         67     Cambridge United     €60K    €1K     1189   
16180   17       47         66      Tranmere Rovers     €60K    €1K     1228   
16181   16       46         66      Tranmere Rovers     €60K    €1K     1321   

       International Reputation  Skill Moves Position Height  Weight    LS  \
0                           5.0          4.0       RF    5'7  159lbs  88+2   
1                           5.0          5.0       ST    6'2  183lbs  91+3   
2                           5.0          5.0       LW    5'9  150lbs  84+3   
3                           4.0          4.0      RCM   5'11  154lbs  82+3   
4                           4.0          4.0       LF    5'8  163lbs  83+3   
...                         ...          ...      ...    ...     ...   ...   
16177                       1.0          2.0       CM    5'9  134lbs  42+2   
16178                       1.0          2.0       ST    6'3  170lbs  45+2   
16179                       1.0          2.0       ST    5'8  148lbs  45+2   
16180                       1.0          2.0       RW   5'10  154lbs  47+2   
16181                       1.0          2.0       CM   5'10  176lbs  43+2   

         ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0      88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1      91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2      84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3      82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   
4      83+3  83+3  89+3  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
16177  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
16178  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2   
16179  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2   
16180  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2   
16181  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2   

         CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0      84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1      81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2      81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3      87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   
4      82+3  82+3  89+3  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
16177  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2  45+2   
16178  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2  32+2  32+2  32+2   
16179  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2  28+2  28+2  28+2   
16180  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2  31+2  31+2  31+2   
16181  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2  47+2  47+2  47+2   

         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0      59+2      84.0       95.0             70.0          90.0     86.0   
1      61+3

## Eliminar NaNs

Comprobamos en que columnas hay valores *NaN*

In [11]:
df.columns[df.isna().any()].tolist()

['Club',
 'International Reputation',
 'Skill Moves',
 'Position',
 'Height',
 'Weight',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

Vemos las filas que contienen aunque sea un *NaN* en cualquiera de sus columnas

In [12]:
df[df.isna().any(axis=1)]

Age  Overall  Potential                            Club   Value   Wage  \
23      26       88         89               FC Bayern München  €69.5M  €315K   
32      30       88         88                           Milan    €57M  €245K   
80      29       85         85  Guangzhou Evergrande Taobao FC    €37M  €235K   
147     24       83         90  Guangzhou Evergrande Taobao FC  €36.5M   €18K   
157     24       83         89                         Chelsea    €35M  €165K   
...    ...      ...        ...                             ...     ...    ...   
16003   21       51         57          Hamilton Academical FC    €50K    €3K   
16004   21       51         60             Guizhou Hengfeng FC    €60K    €2K   
16042   21       50         59             Guizhou Hengfeng FC    €50K    €2K   
16046   20       50         61                  Stabæk Fotball    €40K    €2K   
16062   19       50         65                          Ascoli    €60K    €3K   

       Special  International Reputation  Skill Moves Position Height  Weight  \
23        2171                       4.0          4.0      LAM   5'11  172lbs   
32        1965                       4.0          3.0       LS    6'1  196lbs   
80        2245                       3.0          3.0      LDM    6'0  157lbs   
147       2142                       2.0          4.0      CAM    6'3  176lbs   
157       2115                       3.0          4.0      LCM    5'9  172lbs   
...        ...                       ...          ...      ...    ...     ...   
16003     1277                       1.0          2.0       ST   5'11  165lbs   
16004     1459                       1.0          2.0       CM    6'0  172lbs   
16042     1440                       1.0          2.0       LM   5'10  143lbs   
16046     1278                       1.0          2.0       RB    5'7  143lbs   
16062     1450                       1.0          2.0       CM   5'10  143lbs   

         LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
23     80+3  80+3  80+3  84+3  83+3  83+3  83+3  84+3  85+3  85+3  85+3  83+3   
32     85+3  85+3  85+3  81+3  83+3  83+3  83+3  81+3  80+3  80+3  80+3  78+3   
80     81+3  81+3  81+3  79+3  81+3  81+3  81+3  79+3  81+3  81+3  81+3  79+3   
147    80+2  80+2  80+2  81+2  81+2  81+2  81+2  81+2  81+2  81+2  81+2  80+2   
157    71+3  71+3  71+3  79+3  78+3  78+3  78+3  79+3  81+3  81+3  81+3  80+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
16003  49+2  49+2  49+2  49+2  49+2  49+2  49+2  49+2  47+2  47+2  47+2  47+2   
16004  49+2  49+2  49+2  50+2  50+2  50+2  50+2  50+2  50+2  50+2  50+2  51+2   
16042  44+2  44+2  44+2  47+2  46+2  46+2  46+2  47+2  48+2  48+2  48+2  49+2   
16046  37+2  37+2  37+2  42+2  40+2  40+2  40+2  42+2  39+2  39+2  39+2  43+2   
16062  44+2  44+2  44+2  47+2  45+2  45+2  45+2  47+2  47+2  47+2  47+2  49+2   

        LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
23     81+3  81+3  81+3  83+3  69+3  68+3  68+3  68+3  69+3  65+3  58+3  58+3   
32     71+3  71+3  71+3  78+3  55+3  53+3  53+3  53+3  55+3  51+3  48+3  48+3   
80     82+3  82+3  82+3  79+3  78+3  81+3  81+3  81+3  78+3  78+3  80+3  80+3   
147    79+2  79+2  79+2  80+2  69+2  69+2  69+2  69+2  69+2  67+2  64+2  64+2   
157    82+3  82+3  82+3  80+3  76+3  78+3  78+3  78+3  76+3  74+3  69+3  69+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
16003  40+2  40+2  40+2  47+2  34+2  31+2  31+2  31+2  34+2  33+2  30+2  30+2   
16004  50+2  50+2  50+2  51+2  50+2  50+2  50+2  50+2  50+2  50+2  49+2  49+2   
16042  48+2  48+2  48+2  49+2  50+2  50+2  50+2  50+2  50+2  51+2  50+2  50+2   
16046  38+2  38+2  38+2  43+2  48+2  43+2  43+2  43+2  48+2  48+2  45+2  45+2   
16062  48+2  48+2  48+2  49+2  52+2  51+2  51+2  51+2  52+2  52+2  51+2  51+2   

        RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
23     58+3  65+3      90.0       83.0    

We can get all the rows with the column **Club** equals to NaN. 

In [13]:
df.loc[df.Club != df.Club]

Age  Overall  Potential Club Value Wage  Special  \
399     24       80         85  NaN    €0   €0     2122   
476     33       80         80  NaN    €0   €0     1797   
601     29       79         79  NaN    €0   €0     2038   
775     29       78         78  NaN    €0   €0     1810   
849     30       77         77  NaN    €0   €0     1749   
...    ...      ...        ...  ...   ...  ...      ...   
15338   26       55         58  NaN    €0   €0     1466   
15401   26       55         57  NaN    €0   €0     1366   
15502   23       54         63  NaN    €0   €0     1321   
15577   20       54         67  NaN    €0   €0     1270   
15662   21       53         62  NaN    €0   €0     1247   

       International Reputation  Skill Moves Position Height  Weight    LS  \
399                         2.0          4.0       CM   5'11  165lbs  71+2   
476                         2.0          2.0      LCB    6'4  185lbs  62+2   
601                         2.0          3.0       RB   5'10  154lbs  70+2   
775                         2.0          3.0       ST    6'5  201lbs  77+2   
849                         1.0          2.0       CB    6'2  157lbs  52+2   
...                         ...          ...      ...    ...     ...   ...   
15338                       1.0          2.0      NaN    5'7  159lbs   NaN   
15401                       1.0          2.0       RB    6'4  187lbs  46+2   
15502                       1.0          2.0      NaN    5'9  143lbs   NaN   
15577                       1.0          2.0      NaN    6'0  168lbs   NaN   
15662                       1.0          2.0      NaN    6'3  174lbs   NaN   

         ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
399    71+2  71+2  75+2  75+2  75+2  75+2  75+2  77+2  77+2  77+2  76+2  79+2   
476    62+2  62+2  56+2  58+2  58+2  58+2  56+2  58+2  58+2  58+2  57+2  64+2   
601    70+2  70+2  73+2  72+2  72+2  72+2  73+2  73+2  73+2  73+2  75+2  74+2   
775    77+2  77+2  71+2  74+2  74+2  74+2  71+2  71+2  71+2  71+2  71+2  66+2   
849    52+2  52+2  51+2  51+2  51+2  51+2  51+2  54+2  54+2  54+2  54+2  61+2   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
15338   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
15401  46+2  46+2  45+2  44+2  44+2  44+2  45+2  43+2  43+2  43+2  46+2  43+2   
15502   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
15577   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
15662   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

         CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
399    79+2  79+2  76+2  75+2  77+2  77+2  77+2  75+2  74+2  72+2  72+2  72+2   
476    64+2  64+2  57+2  68+2  74+2  74+2  74+2  68+2  70+2  79+2  79+2  79+2   
601    74+2  74+2  75+2  78+2  75+2  75+2  75+2  78+2  78+2  73+2  73+2  73+2   
775    66+2  66+2  71+2  52+2  52+2  52+2  52+2  52+2  48+2  48+2  48+2  48+2   
849    61+2  61+2  54+2  67+2  72+2  72+2  72+2  67+2  69+2  75+2  75+2  75+2   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
15338   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
15401  43+2  43+2  46+2  51+2  49+2  49+2  49+2  51+2  53+2  57+2  57+2  57+2   
15502   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
15577   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
15662   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
399    74+2      76.0       55.0             60.0          84.0     73.0   
476    70+2      49.0       51.0             81.0          73.0     37.0   
601    78+2      73.0       61.0             69.0          79.0     57.0   
775    48+2      61.0       79.0             86.0          71.0     74.0   
849    69+2      42.0       33.0   

Igual que con los clubs, buscamos en International Reputation cuáles tienen valores *Nan*

In [14]:
df.loc[df["International Reputation"] != df["International Reputation"]]

Age  Overall  Potential                           Club  Value Wage  \
11983   33       62         62                       Rochdale  €120K  €1K   
11984   29       62         62                Boyacá Chicó FC  €300K  €1K   
11985   35       62         62                   Notts County  €140K  €3K   
11986   20       62         72                        Brescia  €425K  €1K   
11987   24       62         66         Hamilton Academical FC  €400K  €1K   
11988   21       62         72                  Śląsk Wrocław  €425K  €1K   
11989   23       62         70         Club Atlético Aldosivi  €450K  €2K   
11990   19       62         78                        Everton  €600K  €5K   
11991   30       62         62                       Hobro IK  €230K  €2K   
11992   21       62         69                   HJK Helsinki  €425K  €1K   
11993   22       62         68                     AS Béziers  €425K  €2K   
11994   28       62         62                 SV Mattersburg  €240K  €3K   
11995   24       62         69                Tranmere Rovers  €375K  €2K   
11996   27       62         62  Shanghai Greenland Shenhua FC  €250K  €3K   
11997   29       62         62               Itagüí Leones FC  €300K  €1K   
11998   34       62         62                      NAC Breda  €150K  €2K   
11999   22       62         70                       Malmö FF  €375K  €1K   
12000   31       62         62                Carlisle United  €200K  €2K   
12001   17       62         82                  VfB Stuttgart  €550K  €2K   
12002   25       62         65         Hamilton Academical FC  €325K  €1K   
12003   26       62         62                      Dundee FC  €325K  €1K   
12004   25       62         65        Suwon Samsung Bluewings  €375K  €2K   
12005   23       62         67                       Al Wehda  €350K  €3K   
12006   27       62         65                   CD Palestino  €300K  €1K   
12007   20       62         69                    Albacete BP  €425K  €1K   
12008   25       62         64                       Al Nassr  €300K  €5K   
12009   24       62         66               TSV 1860 München  €325K  €1K   
12010   25       62         66       FC Admira Wacker Mödling  €325K  €2K   
12011   29       62         62               Grenoble Foot 38  €300K  €1K   
12012   28       62         62                Oldham Athletic  €300K  €3K   
12013   20       62         73                    Hammarby IF  €525K  €1K   
12014   22       62         70               Itagüí Leones FC  €450K  €1K   
12015   37       62         62                  Miedź Legnica  €100K  €1K   
12016   26       62         64            Jaguares de Córdoba  €375K  €1K   
12017   19       62         77                        Bologna  €525K  €1K   
12018   26       62         63                 CD Antofagasta  €290K  €1K   
12019   22       62         70                      Dundee FC  €450K  €1K   
12020   25       62         67                Kristiansund BK  €400K  €1K   
12021   27       62         62               TSV 1860 München  €325K  €2K   
12022   23       62         70                Boyacá Chicó FC  €375K  €1K   
12023   26       62         63                        Al Raed  €290K  €3K   
12024   26       62         62               Deportes Iquique  €325K  €1K   
12025   17       62         79                     FC Utrecht  €550K  €1K   
12026   22       62         69                        Perugia  €350K  €1K   
12027   19       62         77                Montpellier HSC  €650K  €2K   
12028   27       62         62                   Gyeongnam FC  €300K  €1K   
12029   25       62         65    Tiburones Rojos de Veracruz  €375K  €2K   
12030   25       62         66            Guizhou Hengfeng FC  €325K  €2K   

       Special  International Reputation  Skill Moves Position Height Weight  \
11983     1510                       NaN          NaN      NaN    NaN    NaN   
11984     1532                       NaN          NaN      NaN    NaN    

Eliminamos las filas que contienen un *NaN* en la columna International Reputation

In [15]:
df = df.dropna(subset = ['International Reputation'])
df

Age  Overall  Potential                 Club    Value   Wage  Special  \
0       31       94         94         FC Barcelona  €110.5M  €565K     2202   
1       33       94         94             Juventus     €77M  €405K     2228   
2       26       92         93  Paris Saint-Germain  €118.5M  €290K     2143   
3       27       91         92      Manchester City    €102M  €355K     2281   
4       27       91         91              Chelsea     €93M  €340K     2142   
...    ...      ...        ...                  ...      ...    ...      ...   
16177   19       47         65      Crewe Alexandra     €60K    €1K     1307   
16178   19       47         63       Trelleborgs FF     €60K    €1K     1098   
16179   16       47         67     Cambridge United     €60K    €1K     1189   
16180   17       47         66      Tranmere Rovers     €60K    €1K     1228   
16181   16       46         66      Tranmere Rovers     €60K    €1K     1321   

       International Reputation  Skill Moves Position Height  Weight    LS  \
0                           5.0          4.0       RF    5'7  159lbs  88+2   
1                           5.0          5.0       ST    6'2  183lbs  91+3   
2                           5.0          5.0       LW    5'9  150lbs  84+3   
3                           4.0          4.0      RCM   5'11  154lbs  82+3   
4                           4.0          4.0       LF    5'8  163lbs  83+3   
...                         ...          ...      ...    ...     ...   ...   
16177                       1.0          2.0       CM    5'9  134lbs  42+2   
16178                       1.0          2.0       ST    6'3  170lbs  45+2   
16179                       1.0          2.0       ST    5'8  148lbs  45+2   
16180                       1.0          2.0       RW   5'10  154lbs  47+2   
16181                       1.0          2.0       CM   5'10  176lbs  43+2   

         ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0      88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1      91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2      84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3      82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   
4      83+3  83+3  89+3  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
16177  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
16178  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2   
16179  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2   
16180  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2   
16181  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2   

         CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0      84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1      81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2      81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3      87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   
4      82+3  82+3  89+3  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
16177  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2  45+2   
16178  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2  32+2  32+2  32+2   
16179  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2  28+2  28+2  28+2   
16180  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2  31+2  31+2  31+2   
16181  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2  47+2  47+2  47+2   

         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0      59+2      84.0       95.0             70.0          90.0     86.0   
1      61+3

Eliminamos las filas que contengan *NaN* en la columna Club

In [16]:
df = df.dropna(subset = ['Club'])
df

Age  Overall  Potential                 Club    Value   Wage  Special  \
0       31       94         94         FC Barcelona  €110.5M  €565K     2202   
1       33       94         94             Juventus     €77M  €405K     2228   
2       26       92         93  Paris Saint-Germain  €118.5M  €290K     2143   
3       27       91         92      Manchester City    €102M  €355K     2281   
4       27       91         91              Chelsea     €93M  €340K     2142   
...    ...      ...        ...                  ...      ...    ...      ...   
16177   19       47         65      Crewe Alexandra     €60K    €1K     1307   
16178   19       47         63       Trelleborgs FF     €60K    €1K     1098   
16179   16       47         67     Cambridge United     €60K    €1K     1189   
16180   17       47         66      Tranmere Rovers     €60K    €1K     1228   
16181   16       46         66      Tranmere Rovers     €60K    €1K     1321   

       International Reputation  Skill Moves Position Height  Weight    LS  \
0                           5.0          4.0       RF    5'7  159lbs  88+2   
1                           5.0          5.0       ST    6'2  183lbs  91+3   
2                           5.0          5.0       LW    5'9  150lbs  84+3   
3                           4.0          4.0      RCM   5'11  154lbs  82+3   
4                           4.0          4.0       LF    5'8  163lbs  83+3   
...                         ...          ...      ...    ...     ...   ...   
16177                       1.0          2.0       CM    5'9  134lbs  42+2   
16178                       1.0          2.0       ST    6'3  170lbs  45+2   
16179                       1.0          2.0       ST    5'8  148lbs  45+2   
16180                       1.0          2.0       RW   5'10  154lbs  47+2   
16181                       1.0          2.0       CM   5'10  176lbs  43+2   

         ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0      88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1      91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2      84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3      82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   
4      83+3  83+3  89+3  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
16177  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
16178  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2   
16179  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2   
16180  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2   
16181  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2   

         CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0      84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1      81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2      81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3      87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   
4      82+3  82+3  89+3  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
16177  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2  45+2   
16178  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2  32+2  32+2  32+2   
16179  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2  28+2  28+2  28+2   
16180  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2  31+2  31+2  31+2   
16181  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2  47+2  47+2  47+2   

         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0      59+2      84.0       95.0             70.0          90.0     86.0   
1      61+3

Eliminamos la columna Club y la almacenamos en la variable "clb" para posteriormente crear una columna por cada equipo registrado y así poder tener valores numéricos.

In [17]:
clb = df.pop("Club")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)
df.head()

Age  Overall  Potential    Value   Wage  Special  International Reputation  \
0   31       94         94  €110.5M  €565K     2202                       5.0   
1   33       94         94     €77M  €405K     2228                       5.0   
2   26       92         93  €118.5M  €290K     2143                       5.0   
3   27       91         92    €102M  €355K     2281                       4.0   
4   27       91         91     €93M  €340K     2142                       4.0   

   Skill Moves Position Height  Weight    LS    ST    RS    LW    LF    CF  \
0          4.0       RF    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1          5.0       ST    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2          5.0       LW    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3          4.0      RCM   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   
4          4.0       LF    5'8  163lbs  83+3  83+3  83+3  89+3  88+3  88+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   
4  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3  66+3  63+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   
4  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0       84.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             55.0          92.0     82.0       86.0   85.0        83.0   
4             61.0          89.0     80.0       95.0   83.0        79.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         91.0         91.0          78.0         76.0     79.0       91.0   
4         83.0         94.0          94.0         88.0     95.0       90.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     77.0       91.0     63.0     90.0      75.0       91.0        76.0   
4     94.0       82.0     56.0     83.0      66.0       80.0        54.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           61.0         87.0    94.0       79.0       88.0     68.0   
4           41.0         87.0    89.0       86.0       91.0     34.0   

   StandingTackle  SlidingTackle  GKDiving  GKHandling  GKKicking  \
0            28.0           26.0       6.0        11.0       15.0   
1            31.0           23.0       7.0        11.0       15.0   
2            24.0           33.0       9.0         9.0       15.0   
3            58.0           51.0      15.0        13.

De igual forma que hemos hecho con los clubes, lo haremos con las posiciones que puede ocupar un jugador

In [18]:
pos = df.pop("Position")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pos, prefix='pos').reset_index(drop=True)], axis=1, sort=False)
df.head()

Age  Overall  Potential    Value   Wage  Special  International Reputation  \
0   31       94         94  €110.5M  €565K     2202                       5.0   
1   33       94         94     €77M  €405K     2228                       5.0   
2   26       92         93  €118.5M  €290K     2143                       5.0   
3   27       91         92    €102M  €355K     2281                       4.0   
4   27       91         91     €93M  €340K     2142                       4.0   

   Skill Moves Height  Weight    LS    ST    RS    LW    LF    CF    RF    RW  \
0          4.0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2   
1          5.0    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3   
2          5.0    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3   
3          4.0   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3   
4          4.0    5'8  163lbs  83+3  83+3  83+3  89+3  88+3  88+3  88+3  89+3   

    LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM  \
0  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2   
1  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3   
2  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3   
3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3   
4  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3  66+3  63+3  63+3  63+3   

    RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  \
0  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0             70.0   
1  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0             89.0   
2  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0             62.0   
3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0             55.0   
4  66+3  60+3  49+3  49+3  49+3  60+3      81.0       84.0             61.0   

   ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  LongPassing  \
0          90.0     86.0       97.0   93.0        94.0         87.0   
1          81.0     87.0       88.0   81.0        76.0         77.0   
2          84.0     84.0       96.0   88.0        87.0         78.0   
3          92.0     82.0       86.0   85.0        83.0         91.0   
4          89.0     80.0       95.0   83.0        79.0         83.0   

   BallControl  Acceleration  SprintSpeed  Agility  Reactions  Balance  \
0         96.0          91.0         86.0     91.0       95.0     95.0   
1         94.0          89.0         91.0     87.0       96.0     70.0   
2         95.0          94.0         90.0     96.0       94.0     84.0   
3         91.0          78.0         76.0     79.0       91.0     77.0   
4         94.0          94.0         88.0     95.0       90.0     94.0   

   ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0       85.0     68.0     72.0      59.0       94.0        48.0   
1       95.0     95.0     88.0      79.0       93.0        63.0   
2       80.0     61.0     81.0      49.0       82.0        56.0   
3       91.0     63.0     90.0      75.0       91.0        76.0   
4       82.0     56.0     83.0      66.0       80.0        54.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           61.0         87.0    94.0       79.0       88.0     68.0   
4           41.0         87.0    89.0       86.0       91.0     34.0   

   StandingTackle  SlidingTackle  GKDiving  GKHandling  GKKicking  \
0            28.0           26.0       6.0        11.0       15.0   
1            31.0           23.0       7.0        11.0       15.0   
2            24.0           33.0       9.0         9.0       15.0   
3            58.0           51.0      15.0        13.0        5.0   
4            27.0           22.0      

Definimos función para cambiar un precio cómo string a float

In [19]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    
    if(type(x)==str):
        x = x.replace('€', '')
        ret_val = 0.0

        if type(x) == float or type(x) == int:
            ret_val = x
        if 'K' in x:
            if len(x) > 1:
                ret_val = float(x.replace('K', ''))
            ret_val = ret_val *1000
        if 'M' in x:
            if len(x) > 1:
                ret_val = float(x.replace('M', ''))
            ret_val = ret_val * 1000000.0
        return ret_val
    return 0.0

Funcion para cambiar de peso cómo string a float

In [20]:
def weight_to_float(x):
    """
    From weight to Float
    """
    weight = 0.0
    
    if(type(x)==str):
        lbs = x[:3]
        weight = float(lbs)
        return weight
    return 0.0
    

Función para sumar las expectativas en cada posicion de un jugador en string a float

In [21]:
def pos_value_to_float(x):
    """
    From position with + to Float
    """
    pos = 0.0
    if(type(x)==str):
        first = x[:2]
        first = float(first)
        second = x[-1:]
        second = float(second)
        pos = first + second
        return pos
    return 0.0
    

Función para cambiar altura de string a float

In [22]:
def height_to_float(x):
    """
    From height to Float
    """
    
    if(type(x)==str):
        first = x[0]
        first = float(first)
        second = x[2:]
        second = float(second)
        second = second/100
        height = first + second
        return height
    return 0.0
    

Comprobamos los jugadores que tienen *NaN* en la columna 'Release Clause', pero observamos que contienen muchos otros datos que pueden sernos util para el entranamiento por tanto, utilizamos la función que en caso de ser *NaN* le asigne el valor 0

In [23]:
df.loc[df["Release Clause"] != df["Release Clause"]]

Age  Overall  Potential   Value   Wage  Special  \
23      26       88         89  €69.5M  €315K     2171   
32      30       88         88    €57M  €245K     1965   
80      29       85         85    €37M  €235K     2245   
147     24       83         90  €36.5M   €18K     2142   
157     24       83         89    €35M  €165K     2115   
...    ...      ...        ...     ...    ...      ...   
15747   21       51         57    €50K    €3K     1277   
15748   21       51         60    €60K    €2K     1459   
15786   21       50         59    €50K    €2K     1440   
15790   20       50         61    €40K    €2K     1278   
15806   19       50         65    €60K    €3K     1450   

       International Reputation  Skill Moves Height  Weight    LS    ST    RS  \
23                          4.0          4.0   5'11  172lbs  80+3  80+3  80+3   
32                          4.0          3.0    6'1  196lbs  85+3  85+3  85+3   
80                          3.0          3.0    6'0  157lbs  81+3  81+3  81+3   
147                         2.0          4.0    6'3  176lbs  80+2  80+2  80+2   
157                         3.0          4.0    5'9  172lbs  71+3  71+3  71+3   
...                         ...          ...    ...     ...   ...   ...   ...   
15747                       1.0          2.0   5'11  165lbs  49+2  49+2  49+2   
15748                       1.0          2.0    6'0  172lbs  49+2  49+2  49+2   
15786                       1.0          2.0   5'10  143lbs  44+2  44+2  44+2   
15790                       1.0          2.0    5'7  143lbs  37+2  37+2  37+2   
15806                       1.0          2.0   5'10  143lbs  44+2  44+2  44+2   

         LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM  \
23     84+3  83+3  83+3  83+3  84+3  85+3  85+3  85+3  83+3  81+3  81+3  81+3   
32     81+3  83+3  83+3  83+3  81+3  80+3  80+3  80+3  78+3  71+3  71+3  71+3   
80     79+3  81+3  81+3  81+3  79+3  81+3  81+3  81+3  79+3  82+3  82+3  82+3   
147    81+2  81+2  81+2  81+2  81+2  81+2  81+2  81+2  80+2  79+2  79+2  79+2   
157    79+3  78+3  78+3  78+3  79+3  81+3  81+3  81+3  80+3  82+3  82+3  82+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
15747  49+2  49+2  49+2  49+2  49+2  47+2  47+2  47+2  47+2  40+2  40+2  40+2   
15748  50+2  50+2  50+2  50+2  50+2  50+2  50+2  50+2  51+2  50+2  50+2  50+2   
15786  47+2  46+2  46+2  46+2  47+2  48+2  48+2  48+2  49+2  48+2  48+2  48+2   
15790  42+2  40+2  40+2  40+2  42+2  39+2  39+2  39+2  43+2  38+2  38+2  38+2   
15806  47+2  45+2  45+2  45+2  47+2  47+2  47+2  47+2  49+2  48+2  48+2  48+2   

         RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
23     83+3  69+3  68+3  68+3  68+3  69+3  65+3  58+3  58+3  58+3  65+3   
32     78+3  55+3  53+3  53+3  53+3  55+3  51+3  48+3  48+3  48+3  51+3   
80     79+3  78+3  81+3  81+3  81+3  78+3  78+3  80+3  80+3  80+3  78+3   
147    80+2  69+2  69+2  69+2  69+2  69+2  67+2  64+2  64+2  64+2  67+2   
157    80+3  76+3  78+3  78+3  78+3  76+3  74+3  69+3  69+3  69+3  74+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
15747  47+2  34+2  31+2  31+2  31+2  34+2  33+2  30+2  30+2  30+2  33+2   
15748  51+2  50+2  50+2  50+2  50+2  50+2  50+2  49+2  49+2  49+2  50+2   
15786  49+2  50+2  50+2  50+2  50+2  50+2  51+2  50+2  50+2  50+2  51+2   
15790  43+2  48+2  43+2  43+2  43+2  48+2  48+2  45+2  45+2  45+2  48+2   
15806  49+2  52+2  51+2  51+2  51+2  52+2  52+2  51+2  51+2  51+2  52+2   

       Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
23         90.0       83.0             62.0          89.0     90.0       85.0   
32         68.0       92.0             80.0          75.0     90.0       84.0   
80         64.0       81.0             85.0          84.0     75.0       82.0   
147        78.0       80.0             77.0          81.0     76.0       82.0   
157        73.0       62.0             42.0          87.0 

Aplicamos las funciones correspondientes de cada columna para que nos queden solo valores numéricos

In [24]:
df["Value"] = df["Value"].apply(value_to_float)
df["Wage"] = df["Wage"].apply(value_to_float)
df["Release Clause"] = df["Release Clause"].apply(value_to_float)
df["Weight"] = df["Weight"].apply(weight_to_float)
df["Height"] = df["Height"].apply(height_to_float)
df

Age  Overall  Potential        Value      Wage  Special  \
0       31       94         94  110500000.0  565000.0     2202   
1       33       94         94   77000000.0  405000.0     2228   
2       26       92         93  118500000.0  290000.0     2143   
3       27       91         92  102000000.0  355000.0     2281   
4       27       91         91   93000000.0  340000.0     2142   
...    ...      ...        ...          ...       ...      ...   
15921   19       47         65      60000.0    1000.0     1307   
15922   19       47         63      60000.0    1000.0     1098   
15923   16       47         67      60000.0    1000.0     1189   
15924   17       47         66      60000.0    1000.0     1228   
15925   16       46         66      60000.0    1000.0     1321   

       International Reputation  Skill Moves  Height  Weight    LS    ST  \
0                           5.0          4.0    5.07   159.0  88+2  88+2   
1                           5.0          5.0    6.02   183.0  91+3  91+3   
2                           5.0          5.0    5.09   150.0  84+3  84+3   
3                           4.0          4.0    5.11   154.0  82+3  82+3   
4                           4.0          4.0    5.08   163.0  83+3  83+3   
...                         ...          ...     ...     ...   ...   ...   
15921                       1.0          2.0    5.09   134.0  42+2  42+2   
15922                       1.0          2.0    6.03   170.0  45+2  45+2   
15923                       1.0          2.0    5.08   148.0  45+2  45+2   
15924                       1.0          2.0    5.10   154.0  47+2  47+2   
15925                       1.0          2.0    5.10   176.0  43+2  43+2   

         RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0      88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1      91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2      84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3      82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   
4      83+3  89+3  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
15921  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2   
15922  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2  35+2   
15923  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2  38+2   
15924  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2  39+2   
15925  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2  45+2   

        RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0      84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1      81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2      81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3      87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   
4      82+3  89+3  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
15921  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2  45+2  45+2   
15922  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2  32+2  32+2  32+2  29+2   
15923  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2  28+2  28+2  28+2  33+2   
15924  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2  31+2  31+2  31+2  35+2   
15925  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2  47+2  47+2  47+2  46+2   

       Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0          84.0       95.0             70.0          90.0     86.0       97.0   
1          84.0       94.0             89.0          81.0     87.0       88.0   
2          79.0       87.0             62.0          84.0     84.0       96.0   
3          93.0       82.0     

Aplicamos la función posicón a float para cada una de las proyecciones por posición de cada jugador

In [25]:
df["LS"] = df["LS"].apply(pos_value_to_float)
df["ST"] = df["ST"].apply(pos_value_to_float)
df["RS"] = df["RS"].apply(pos_value_to_float)
df["LW"] = df["LW"].apply(pos_value_to_float)
df["LF"] = df["LF"].apply(pos_value_to_float)
df["CF"] = df["CF"].apply(pos_value_to_float)
df["RF"] = df["RF"].apply(pos_value_to_float)
df["RW"] = df["RW"].apply(pos_value_to_float)
df["LAM"] = df["LAM"].apply(pos_value_to_float)
df["CAM"] = df["CAM"].apply(pos_value_to_float)
df["RAM"] = df["RAM"].apply(pos_value_to_float)
df["LM"] = df["LM"].apply(pos_value_to_float)
df["LCM"] = df["LCM"].apply(pos_value_to_float)
df["CM"] = df["CM"].apply(pos_value_to_float)
df["RCM"] = df["RCM"].apply(pos_value_to_float)
df["RM"] = df["RM"].apply(pos_value_to_float)
df["LWB"] = df["LWB"].apply(pos_value_to_float)
df["LDM"] = df["LDM"].apply(pos_value_to_float)
df["CDM"] = df["CDM"].apply(pos_value_to_float)
df["RDM"] = df["RDM"].apply(pos_value_to_float)
df["RWB"] = df["RWB"].apply(pos_value_to_float)
df["LB"] = df["LB"].apply(pos_value_to_float)
df["LCB"] = df["LCB"].apply(pos_value_to_float)
df["CB"] = df["CB"].apply(pos_value_to_float)
df["RCB"] = df["RCB"].apply(pos_value_to_float)
df["RB"] = df["RB"].apply(pos_value_to_float)

df

Age  Overall  Potential        Value      Wage  Special  \
0       31       94         94  110500000.0  565000.0     2202   
1       33       94         94   77000000.0  405000.0     2228   
2       26       92         93  118500000.0  290000.0     2143   
3       27       91         92  102000000.0  355000.0     2281   
4       27       91         91   93000000.0  340000.0     2142   
...    ...      ...        ...          ...       ...      ...   
15921   19       47         65      60000.0    1000.0     1307   
15922   19       47         63      60000.0    1000.0     1098   
15923   16       47         67      60000.0    1000.0     1189   
15924   17       47         66      60000.0    1000.0     1228   
15925   16       46         66      60000.0    1000.0     1321   

       International Reputation  Skill Moves  Height  Weight    LS    ST  \
0                           5.0          4.0    5.07   159.0  90.0  90.0   
1                           5.0          5.0    6.02   183.0  94.0  94.0   
2                           5.0          5.0    5.09   150.0  87.0  87.0   
3                           4.0          4.0    5.11   154.0  85.0  85.0   
4                           4.0          4.0    5.08   163.0  86.0  86.0   
...                         ...          ...     ...     ...   ...   ...   
15921                       1.0          2.0    5.09   134.0  44.0  44.0   
15922                       1.0          2.0    6.03   170.0  47.0  47.0   
15923                       1.0          2.0    5.08   148.0  47.0  47.0   
15924                       1.0          2.0    5.10   154.0  49.0  49.0   
15925                       1.0          2.0    5.10   176.0  45.0  45.0   

         RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0      90.0  94.0  95.0  95.0  95.0  94.0  95.0  95.0  95.0  93.0  86.0  86.0   
1      94.0  92.0  93.0  93.0  93.0  92.0  91.0  91.0  91.0  91.0  84.0  84.0   
2      87.0  92.0  92.0  92.0  92.0  92.0  92.0  92.0  92.0  91.0  84.0  84.0   
3      85.0  90.0  90.0  90.0  90.0  90.0  91.0  91.0  91.0  91.0  90.0  90.0   
4      86.0  92.0  91.0  91.0  91.0  92.0  92.0  92.0  92.0  92.0  85.0  85.0   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
15921  44.0  46.0  46.0  46.0  46.0  46.0  47.0  47.0  47.0  46.0  47.0  47.0   
15922  47.0  41.0  44.0  44.0  44.0  41.0  42.0  42.0  42.0  40.0  37.0  37.0   
15923  47.0  47.0  48.0  48.0  48.0  47.0  46.0  46.0  46.0  46.0  40.0  40.0   
15924  49.0  49.0  48.0  48.0  48.0  49.0  47.0  47.0  47.0  48.0  41.0  41.0   
15925  45.0  47.0  46.0  46.0  46.0  47.0  47.0  47.0  47.0  48.0  47.0  47.0   

        RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0      86.0  93.0  66.0  63.0  63.0  63.0  66.0  61.0  49.0  49.0  49.0  61.0   
1      84.0  91.0  68.0  64.0  64.0  64.0  68.0  64.0  56.0  56.0  56.0  64.0   
2      84.0  91.0  68.0  63.0  63.0  63.0  68.0  63.0  50.0  50.0  50.0  63.0   
3      90.0  91.0  80.0  80.0  80.0  80.0  80.0  76.0  69.0  69.0  69.0  76.0   
4      85.0  92.0  69.0  66.0  66.0  66.0  69.0  63.0  52.0  52.0  52.0  63.0   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
15921  47.0  46.0  46.0  47.0  47.0  47.0  46.0  47.0  47.0  47.0  47.0  47.0   
15922  37.0  40.0  32.0  33.0  33.0  33.0  32.0  31.0  34.0  34.0  34.0  31.0   
15923  40.0  46.0  36.0  32.0  32.0  32.0  36.0  35.0  30.0  30.0  30.0  35.0   
15924  41.0  48.0  38.0  34.0  34.0  34.0  38.0  37.0  33.0  33.0  33.0  37.0   
15925  47.0  48.0  48.0  48.0  48.0  48.0  48.0  48.0  49.0  49.0  49.0  48.0   

       Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0          84.0       95.0             70.0          90.0     86.0       97.0   
1          84.0       94.0             89.0          81.0     87.0       88.0   
2          79.0       87.0             62.0          84.0     84.0       96.0   
3          93.0       82.0     

## Prediction

Finally we want to try to predict the **value**

In [26]:
val = df.pop("Value")

At the same time we have the data we want to predict in a numeric format, the value information. So now we can use this data to train a model. But first we need to split the data to being able to known its true performance. Utilizamos un 20% tests y 80% predicción de entrenamiento

In [27]:
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.2, random_state=42)

In [28]:
len(X_train)

12740

Now we train a LinearRegression model.

In [29]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

Finally we get a metric $R^2$ for the regression, we use the implementation from <a href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html"> sickit-learn</a>.

In [30]:
preds = reg.predict(X_test)

Tenemos la predicción del primer jugador

In [31]:
preds[0]

623596.5106220534

Test del primer jugador

In [32]:
y_test[0]

110500000.0

In [33]:
r2_score(preds, y_test)

0.9742369100984668

Finalmente obtenemos un error cuadrático medio que obtenemos es del 0.9742369100984668, por tanto podemos finalizar que hemos obtenido un buen resultado, ya que, está muy próximo a 1.